provider/coredns.go

```diff
@@ -83,52 +90,56 @@ type Service struct {
 }
 
 type etcdClient struct {
-	api etcd.KeysAPI
+	client *etcdcv3.Client
+	ctx    context.Context
 }
 
-var _ skyDNSClient = etcdClient{}
+var _ coreDNSClient = etcdClient{}
 
 // GetService return all Service records stored in etcd stored anywhere under the given key (recursively)
 func (c etcdClient) GetServices(prefix string) ([]*Service, error) {
-	var result []*Service
-	opts := &etcd.GetOptions{Recursive: true}
-	data, err := c.api.Get(context.Background(), prefix, opts)
+	ctx, cancel := context.WithTimeout(c.ctx, etcdTimeout)
+	defer cancel()
+
+	path := prefix
+	r, err := c.client.Get(ctx, path, etcdcv3.WithPrefix())
 	if err != nil {
-		if etcd.IsKeyNotFound(err) {
-			return nil, nil
-		}
 		return nil, err
 	}
 
-	queue := list.New()
-	queue.PushFront(data.Node)
-	for queueNode := queue.Front(); queueNode != nil; queueNode = queueNode.Next() {
-		node := queueNode.Value.(*etcd.Node)
-		if node.Dir {
-			for _, childNode := range node.Nodes {
-				queue.PushBack(childNode)
-			}
-			continue
+	var sx []*Service
+	bx := make(map[Service]bool)
+	for _, n := range r.Kvs {
+		serv := new(Service)
```

Comment:
wouldn't it be better to name it `s`or `svc`. `serv`sounds more like server to me 😄

Reply:
updated.